pkg/controller/common.go

```diff
@@ -636,331 +649,793 @@ func (c *SharedState) isVolumeVisited(azVolumeName string) bool {
+	pvs := make([]*v1.PersistentVolume, len(volumes))
+	for i, volume := range volumes {
+		azVolume, err := azureutils.GetAzVolume(ctx, c.cachedClient, c.azClient, volume, c.objectNamespace, true)
 		if err != nil {
+			klog.V(5).Infof("AzVolume for volume %s is not found.", volume)
 			return nil, err
 		}
-		nodes = append(nodes, selected...)
+
+		var pv v1.PersistentVolume
+		if err := c.cachedClient.Get(ctx, types.NamespacedName{Name: azVolume.Status.PersistentVolume}, &pv); err != nil {
+			return nil, err
+		}
+		pvs[i] = &pv
 	}
 
-	klog.Infof("Selected nodes (%+v) for replica AzVolumeAttachments for volumes (%+v)", nodes, volumes)
-	return nodes[:min(len(nodes), numReplica)], nil
-}
+	var filterPlugins = []filterPlugin{
+		&interPodAffinityFilter{},
+		&interPodAntiAffinityFilter{},
+		&podTolerationFilter{},
+		&podNodeAffinityFilter{},
+		&podNodeSelectorFilter{},
+		&volumeNodeSelectorFilter{},
+	}
 
-func getSupportedZones(nodeSelectorTerms []v1.NodeSelectorTerm, topologyKey string) set {
-	// Get the list of supported zones for pv
-	supportedZones := set{}
-	if len(nodeSelectorTerms) > 0 {
-		for _, term := range nodeSelectorTerms {
-			if len(term.MatchExpressions) > 0 {
-				for _, matchExpr := range term.MatchExpressions {
-					if matchExpr.Key == topologyKey {
-						for _, value := range matchExpr.Values {
-							if value != "" && !supportedZones.has(value) {
-								supportedZones.add(value)
-							}
-						}
-					}
-				}
+	filteredNodes := nodes
+	for _, filterPlugin := range filterPlugins {
+		filterPlugin.setup(pods, pvs, c)
+		if updatedFilteredNodes, err := filterPlugin.filter(ctx, filteredNodes); err != nil {
+			klog.Errorf("failed to filter node with filter plugin (%s): %v", filterPlugin.name(), err)
+		} else {
+			filteredNodes = updatedFilteredNodes
+			nodeStrs := make([]string, len(filteredNodes))
+			for i, filteredNode := range filteredNodes {
+				nodeStrs[i] = filteredNode.Name
 			}
+			klog.Infof("Filtered node list from filter plugin (%s): %+v", filterPlugin.name(), nodeStrs)
 		}
 	}
-	return supportedZones
+
+	return filteredNodes, nil
 }
 
-func getNodeRemainingCapacity(ctx context.Context, cachedClient client.Client, nodeObj *v1.Node, nodeName string) (int, error) {
-	// get node if necessary
-	if nodeObj == nil {
-		nodeObj = &v1.Node{}
-		if err := cachedClient.Get(ctx, types.NamespacedName{Name: nodeName}, nodeObj); err != nil {
-			return -1, err
-		}
+func (c *SharedState) prioritizeNodes(ctx context.Context, pods []v1.Pod, volumes []string, nodes []v1.Node) []v1.Node {
+	nodeScores := map[string]int{}
+	for _, node := range nodes {
+		nodeScores[node.Name] = 0
 	}
 
-	// get all replica azvolumeattachments on the node
-	attachments, err := getAzVolumeAttachmentsForNode(ctx, cachedClient, nodeName, all)
-	if err != nil {
-		return -1, err
+	var nodeScorerPlugins = []nodeScorerPlugin{
+		&scoreByNodeCapacity{},
+		&scoreByReplicaCount{},
```

Comment:
By scoring the nodes 1st by node capacity and then by the number of replica attachments, isn't there a chance that we end up attaching the replicas for 2 volumes from the same pod on 2 different nodes, even though there is enough capacity on node that already has the replica attachments from the other volumes of the pod?

I am wondering if it should be done this way - sort the nodes based on the number of existing replica attachments that they have from the other volumes of the pod and then loop over the sorted list to filter out the nodes that have enough remaining capacity? Similar to what is there in the master branch.

Reply:
Yes, that was actually the motivation behind the addition of coefficients in scoring method. We intend to score the nodes with replica mounts with a higher coefficient so that those nodes get prioritized over the nodes with higher capacity but less replica mounts. However, that said, I do think the currently assigned coefficients is too low and leaves room for the abovementioned scenario to occur.

I do think we still need scoring based on capacity as well for cases where the number of replica mounts between the nodes are the same. On that note, I suggest that we increase the scoring coefficient for nodes with replica mounts.

https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/872578519a593c3de2f6671c65490b501686aa16/pkg/controller/common.go#L1164
https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/872578519a593c3de2f6671c65490b501686aa16/pkg/controller/common.go#L1197

Reply:
If we increase the scoring coefficient for nodes with replica mounts, say we make it 10, the nodes that have remaining capacity of 0(because of already attached replica attachments), might get scored higher than the nodes that have capacity. 

Is it possible to:
1. Score the nodes based on replica count 1st
2. Then, score the nodes based on remaining capacity, while doing that - check if the remaining capacity is less than what's required, discard that node by putting a -ve score to that node?

Reply:
> If we increase the scoring coefficient for nodes with replica mounts, say we make it 10, the nodes that have remaining capacity of 0(because of already attached replica attachments), might get scored higher than the nodes that have capacity.

This scenario should not be possible because we filter out nodes without sufficient capacity.

https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/872578519a593c3de2f6671c65490b501686aa16/pkg/controller/common.go#L1160-L1161

Reply:
Ah, that makes sense! I guess in that case having a large nodeScoreHighCoefficient  will solve it. 
On the same note, do we  need a nodeScoreLowCoefficient? We could just nodeScores[nodeName] = score + remainingCapacity ?

Reply:
Yes, since `nodeScoreLowCoefficient = 1`, it is trivial to add the coefficient to the equation. It was added for consistency reasons but it can be removed if it seems too trivial.